sui_programmability/adapter/src/adapter.rs

```diff
@@ -800,6 +807,69 @@ pub fn resolve_and_type_check(
     })
 }
 
+/// Check that for each pair of a shared object and a descendant of it (through object ownership),
+/// at least one of the types of the shared object and the descendant must be defined in the
+/// same module as the function being called (somewhat similar to Rust's orphan rule).
+fn check_child_object_of_shared_object(
+    objects: &BTreeMap<ObjectID, impl Borrow<Object>>,
+    object_type_map: &BTreeMap<ObjectID, ModuleId>,
+    current_module: ModuleId,
+) -> SuiResult {
+    // ancestor_map is a cache that remembers the top ancestor of each object.
+    // Top ancestor is the object at the top in the object ownership chain whose
+    // parent is no longer an object.
+    let mut ancestor_map: BTreeMap<ObjectID, ObjectID> = BTreeMap::new();
+    for (object_id, obj) in objects {
+        // We only need to look at objects that are owned by other objects.
+        if !matches!(obj.borrow().owner, Owner::ObjectOwner(_)) {
+            continue;
+        }
+        // We trace the object up through the ownership chain, until we either hit
+        // the top (no more parent object), or we hit an object that's already
+        // in the `ancestor_map` (because we visited this object previously).
+        // We then have a pair between this object and its top ancestor. If the top
+        // ancestor is a shared object, we check on their types.
+        let mut ancestor_stack = vec![];
+        let mut cur_id = *object_id;
+        let ancestor_id = loop {
+            if let Some(ancestor) = ancestor_map.get(&cur_id) {
+                break *ancestor;
+            }
+            ancestor_stack.push(cur_id);
+            let owner = objects.get(&cur_id).unwrap().borrow().owner;
+            match owner {
+                Owner::ObjectOwner(parent_id) => {
+                    cur_id = parent_id.into();
+                }
+                _ => {
```

Comment:
Exhaustive pattern matching here might be useful for future-proofing if/when we extend the authentication language later.